gravitee-apim-gateway/gravitee-apim-gateway-handlers/gravitee-apim-gateway-handlers-api/src/main/java/io/gravitee/gateway/jupiter/handlers/api/v4/DefaultApiReactor.java

```diff
@@ -163,43 +183,77 @@ private void prepareContextAttributes(MutableExecutionContext ctx) {
         ctx.setAttribute(ContextAttributes.ATTR_ORGANIZATION, api.getOrganizationId());
         ctx.setAttribute(ContextAttributes.ATTR_ENVIRONMENT, api.getEnvironmentId());
         ctx.setInternalAttribute(ContextAttributes.ATTR_API, api);
-        ctx.setInternalAttribute(InternalContextAttributes.ATTR_INTERNAL_INVOKER, defaultInvoker);
+        ctx.setInternalAttribute(ATTR_INTERNAL_INVOKER, defaultInvoker);
+        ctx.setInternalAttribute(LoggingContext.ATTR_INTERNAL_LOGGING_CONTEXT, loggingContext);
+    }
+
+    private void prepareMetrics(HttpExecutionContext ctx) {
+        final GenericRequest request = ctx.request();
+        final Metrics metrics = request.metrics();
+
+        metrics.setApi(api.getId());
+        metrics.setPath(request.pathInfo());
     }
 
     private Completable handleRequest(final MutableExecutionContext ctx) {
-        return platformFlowChain
-            .execute(ctx, REQUEST)
-            .andThen(securityChain.execute(ctx))
-            .andThen(handleEntrypointRequest(ctx))
-            .andThen(platformFlowChain.execute(ctx, MESSAGE_REQUEST))
-            .andThen(executeProcessorsChain(ctx, apiPreProcessorChain, REQUEST))
-            .andThen(apiPlanFlowChain.execute(ctx, REQUEST))
-            .andThen(apiPlanFlowChain.execute(ctx, MESSAGE_REQUEST))
-            .andThen(apiFlowChain.execute(ctx, REQUEST))
-            .andThen(apiFlowChain.execute(ctx, MESSAGE_REQUEST))
-            .andThen(invokeBackend(ctx))
-            .andThen(apiPlanFlowChain.execute(ctx, RESPONSE))
-            .andThen(apiPlanFlowChain.execute(ctx, MESSAGE_RESPONSE))
-            .andThen(apiFlowChain.execute(ctx, RESPONSE))
-            .andThen(apiFlowChain.execute(ctx, MESSAGE_RESPONSE))
-            .andThen(platformFlowChain.execute(ctx, RESPONSE))
-            .andThen(platformFlowChain.execute(ctx, MESSAGE_RESPONSE))
-            .andThen(executeProcessorsChain(ctx, apiMessageProcessorChain, MESSAGE_RESPONSE))
-            .andThen(handleEntrypointResponse(ctx))
-            .andThen(executeProcessorsChain(ctx, apiPostProcessorChain, RESPONSE))
-            .onErrorResumeNext(error -> processThrowable(ctx, error))
+        // Execute platform flow chain.
+        return new CompletableReactorChain(platformFlowChain.execute(ctx, REQUEST))
+            // Resolve entrypoint and prepare request to be handled.
+            .chainWith(handleEntrypointRequest(ctx))
```

Comment:
You're right. I see no real issue for most of the entrypoint except for websocket because, in that case, the connection will be upgraded before the security constraints are verified.
I'll will try to find a way to execute the security chain before and see if it does the job for all the case (sync and async).

Good catch 👍